src/swapRouter.ts

```diff
@@ -300,14 +336,85 @@ export abstract class SwapRouter {
     return {
       calldata: MulticallExtended.encodeMulticall(calldatas, options.deadlineOrPreviousBlockhash),
-      value: toHex(totalValue.quotient),
+      value: toHex(inputIsNative ? totalAmountIn.quotient : ZERO),
+    }
+  }
+
+  /**
+   * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
+   * @param trade to produce call parameters for
+   * @param options options for the call parameters
+   */
+  public static swapAndAddCallParameters(
+    trades:
+      | Trade<Currency, Currency, TradeType>
+      | V2Trade<Currency, Currency, TradeType>
+      | V3Trade<Currency, Currency, TradeType>
+      | (V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>)[],
+    options: SwapOptions,
+    tokenInBalanceDesired: CurrencyAmount<Currency>,
+    tokenOutBalanceDesired: CurrencyAmount<Currency>,
+    position: Position,
+    addLiquidityOptions: AddLiquidityOptions
+  ): MethodParameters {
+    const { calldatas, inputIsNative, outputIsNative, totalAmountIn, totalAmountOut } = SwapRouter.encodeSwaps(
+      trades,
+      options
+    )
+
+    // if tokens are native they will be converted to WETH9
+    const tokenIn  = inputIsNative ? WETH9[1].address : tokenInBalanceDesired.currency.wrapped
+    const tokenOut = outputIsNative ? WETH9[1].address : tokenOutBalanceDesired.currency.wrapped
+
+    // if swap output does not make up whole outputTokenBalanceDesired, pull in remaining tokens for adding liquidity
+    const amountOutRemainingBalance = tokenOutBalanceDesired.subtract(totalAmountOut)
+    if (amountOutRemainingBalance.greaterThan(CurrencyAmount.fromRawAmount(tokenOutBalanceDesired.currency, 0))) {
+      // if output is native, convert value to WETH9, else pull ERC20 token
+      outputIsNative
+        ? calldatas.push(
+            SwapRouter.INTERFACE.encodeFunctionData('pay', [
+              WETH9[1].address,
+              SWAP_ROUTER_ADDRESS,
+              SWAP_ROUTER_ADDRESS,
+              amountOutRemainingBalance.quotient,
+            ])
+          )
+        : calldatas.push(PaymentsExtended.encodePull(tokenOut, amountOutRemainingBalance.quotient))
+    }
+
+    // if input is native, convert to WETH9, else pull ERC20 token
+    inputIsNative
+      ? calldatas.push(
+          SwapRouter.INTERFACE.encodeFunctionData('pay', [
+            WETH9[1].address,
+            SWAP_ROUTER_ADDRESS,
+            SWAP_ROUTER_ADDRESS,
+            tokenInBalanceDesired.quotient,
+          ])
+        )
+      : calldatas.push(PaymentsExtended.encodePull(tokenIn, tokenInBalanceDesired.quotient))
+
+    // approve token balances to NFTManager
+    calldatas.push(ApproveAndCall.encodeApproveMax(tokenIn))
+    calldatas.push(ApproveAndCall.encodeApproveMax(tokenOut))
+
+    // encode NFTManager add liquidity
+    calldatas.push(NonfungiblePositionManager.addCallParameters(position, addLiquidityOptions).calldata)
+
+    // sweep remaining tokens
```

Comment:
if one of these tokens were native, we are returning WETH9 atm. can change this.

Reply:
yeah we should always refund in the same currency